File: src/lib/providers/impl/bodian.ts
import axios from 'axios';
import { MusicItem, MusicProvider, PlayInfo } from '@/types/music';

const SEARCH_API_URL = 'https://bd-api.kuwo.cn/api/search/music/list';
const CGG_API_URL = 'https://kw-api.cenguigui.cn/';
const TIANBAO_API_URL = 'https://mobi.kuwo.cn/mobi.s';
const REQUEST_TIMEOUT = 20000;

const SEARCH_HEADERS = {
  'user-agent': 'Dart/3.3 (dart:io)',
  'plat': 'win',
  'accept-encoding': 'gzip',
  'api-ver': 'application/json',
  'channel': 'W1',
  'brand': 'Windows 11 Pro for Workstations',
  'net': 'wifi',
  'content-type': 'application/json',
  'ver': '1.1.5',
  'svrver': '13',
  'devid': 'coco-bodian',
  'qimei36': 'coco-bodian',
};

type BodianSearchItem = {
  id?: string | number;
  name?: string;
  artist?: string;
  album?: string;
  albumPic?: string;
  freeSign?: string;
  fsig?: string;
};

function extractExt(url: string, fallback = 'mp3') {
  const clean = url.split('?')[0];
  const parts = clean.split('.');
  return parts.length > 1 ? parts[parts.length - 1] : fallback;
}

function normalizeCover(value?: string) {
  return value || undefined;
}

export class BodianProvider implements MusicProvider {
  name = 'bodian';

  async search(query: string): Promise<MusicItem[]> {
    try {
      const { data } = await axios.get(SEARCH_API_URL, {
        headers: SEARCH_HEADERS,
        params: {
          pn: '0',
          rn: '10',
          keyword: query.trim(),
          correct: '1',
          uid: '-1',
          token: '',
        },
        timeout: REQUEST_TIMEOUT,
      });

      const list = (((data || {}).data || {}).resultList || []) as BodianSearchItem[];
      return list
        .map((item) => ({
          id: String(item.id || ''),
          title: String(item.name || '').trim(),
          artist: item.artist || '未知歌手',
          album: item.album || undefined,
          cover: normalizeCover(item.albumPic),
          provider: this.name,
          extra: {
            albumPic: item.albumPic || undefined,
            freeSign: item.freeSign || item.fsig || undefined,
          },
        }))
        .filter((item) => item.id && item.title);
    } catch (error) {
      console.error('Bodian search error:', error);
      return [];
    }
  }

  async getPlayInfo(id: string, extra?: unknown): Promise<PlayInfo> {
    try {
      const fallbackCover = this.extractCover(extra);
      try {
        const info = await this.getByCenguigui(id);
        return {
          url: info.url,
          type: extractExt(info.url, 'flac'),
          cover: info.cover || fallbackCover,
          bitrate: info.bitrate,
        };
      } catch {
        const info = await this.getByTianbao(id);
        return {
          url: info.url,
          type: extractExt(info.url, 'flac'),
          cover: fallbackCover,
          bitrate: info.bitrate,
        };
      }
    } catch (error) {
      console.error('Bodian getPlayInfo error:', error);
      throw error;
    }
  }

  private extractCover(extra: unknown) {
    const payload = extra as { albumPic?: string } | undefined;
    return payload?.albumPic || undefined;
  }

  private async getByCenguigui(id: string) {
    const { data } = await axios.get(CGG_API_URL, {
      params: {
        id,
        type: 'song',
        level: 'lossless',
        format: 'json',
      },
      timeout: REQUEST_TIMEOUT,
    });

    const payload = (data || {}).data || {};
    const url = String(payload.url || '').trim();
    if (!url.startsWith('http')) {
      throw new Error('Invalid cenguigui url');
    }

    return {
      url,
      cover: payload.pic ? String(payload.pic) : undefined,
      bitrate: 'lossless',
    };
  }

  private async getByTianbao(id: string) {
    const { data } = await axios.get(TIANBAO_API_URL, {
      headers: {
        'User-Agent': 'Dart/2.19 (dart:io)',
        'plat': 'ar',
        'channel': 'aliopen',
      },
      params: {
        f: 'web',
        user: '2333333',
        source: 'kwplayerhd_ar_4.3.0.8_tianbao_T1A_qirui.apk',
        type: 'convert_url_with_sign',
        br: '2000kflac',
        rid: id,
      },
      timeout: REQUEST_TIMEOUT,
    });

    const payload = (data || {}).data || {};
    const url = String(payload.url || '').trim();
    if (!url.startsWith('http')) {
      throw new Error('Invalid tianbao url');
    }

    return {
      url,
      bitrate: '2000kflac',
    };
  }
}

File: desktop/app/services/providers/bodian.py
# coding: utf-8
import base64
import logging
from typing import Any

from requests import RequestException

from app.models.music import LyricData, MusicItem, PlayInfo
from app.services.errors import ProviderNetworkError

from .base import MusicProvider
from .http_client import ProviderHttpClient
from .utils import clean_lyric, clean_text, extract_ext, is_http_url, parse_lrc_lines

LOGGER = logging.getLogger(__name__)
SEARCH_API_URL = "https://bd-api.kuwo.cn/api/search/music/list"
CGG_API_URL = "https://kw-api.cenguigui.cn/"
TIANBAO_API_URL = "https://mobi.kuwo.cn/mobi.s"
KUWO_LYRIC_URL = "http://mlyric.kuwo.cn/mobi.s"
REQUEST_TIMEOUT = 20
DEFAULT_PLAY_LEVEL = "standard"
DEFAULT_PLAY_BR = "320kmp3"

SEARCH_HEADERS = {
    "user-agent": "Dart/3.3 (dart:io)",
    "plat": "win",
    "accept-encoding": "gzip",
    "api-ver": "application/json",
    "channel": "W1",
    "brand": "Windows 11 Pro for Workstations",
    "net": "wifi",
    "content-type": "application/json",
    "ver": "1.1.5",
    "svrver": "13",
    "devid": "coco-bodian",
    "qimei36": "coco-bodian",
}


class BodianProvider(MusicProvider):
    name = "bodian"

    def __init__(self) -> None:
        self._http = ProviderHttpClient()

    def search(self, query: str, limit: int = 20, offset: int = 0) -> list[MusicItem]:
        try:
            data = self._http.get_json(
                SEARCH_API_URL,
                headers=SEARCH_HEADERS,
                params={
                    "pn": "0",
                    "rn": "10",
                    "keyword": query.strip(),
                    "correct": "1",
                    "uid": "-1",
                    "token": "",
                },
                timeout=REQUEST_TIMEOUT,
            )
        except RequestException:
            LOGGER.exception("Bodian search error")
            return []

        payload = data.get("data", {}) if isinstance(data, dict) else {}
        items = payload.get("resultList", []) if isinstance(payload, dict) else []
        if not isinstance(items, list):
            return []
        return [item for item in (self._map_item(raw_item) for raw_item in items) if item]

    def get_play_info(self, song_id: str, extra: dict[str, Any] | None = None) -> PlayInfo:
        fallback_cover = extra.get("albumPic") if extra and isinstance(extra.get("albumPic"), str) else None
        try:
            info = self._get_by_cenguigui(song_id)
            return PlayInfo(
                url=info["url"],
                type=extract_ext(info["url"], "mp3"),
                bitrate=info.get("bitrate"),
                cover=info.get("cover") or fallback_cover,
            )
        except (RequestException, ValueError):
            info = self._get_by_tianbao(song_id)
            return PlayInfo(
                url=info["url"],
                type=extract_ext(info["url"], "mp3"),
                bitrate=info.get("bitrate"),
                cover=fallback_cover,
            )

    def get_lyric(self, song_id: str, extra: dict[str, Any] | None = None) -> LyricData:
        try:
            lyric = clean_lyric(self._get_lyric_by_official_api(song_id))
        except (ProviderNetworkError, RequestException, ValueError):
            info = self._get_by_cenguigui(song_id)
            lyric = clean_lyric(info.get("lyric", ""))

        if "歌词获取失败" in lyric:
            lyric = ""
        return LyricData(songid=song_id, provider=self.name, lines=parse_lrc_lines(lyric), lrc=lyric)

    def _map_item(self, item: Any) -> MusicItem | None:
        if not isinstance(item, dict):
            return None
        song_id = str(item.get("id") or "")
        title = clean_text(str(item.get("name") or ""))
        if not song_id or not title:
            return None
        album_pic = item.get("albumPic") if isinstance(item.get("albumPic"), str) else None
        free_sign = item.get("freeSign") or item.get("fsig")
        return MusicItem(
            id=song_id,
            title=title,
            artist=item.get("artist") or "未知歌手",
            album=item.get("album") or None,
            cover=album_pic,
            provider=self.name,
            extra={"albumPic": album_pic, "freeSign": free_sign or None},
        )

    def _get_by_cenguigui(self, song_id: str) -> dict[str, str]:
        data = self._http.get_json(
            CGG_API_URL,
            params={"id": song_id, "type": "song", "level": DEFAULT_PLAY_LEVEL, "format": "json"},
            timeout=REQUEST_TIMEOUT,
        )
        payload = data.get("data", {}) if isinstance(data, dict) else {}
        url = clean_text(str(payload.get("url", ""))) if isinstance(payload, dict) else ""
        if not is_http_url(url):
            raise ValueError("Invalid cenguigui url")
        cover = payload.get("pic") if isinstance(payload.get("pic"), str) else None
        lyric = payload.get("lyric") if isinstance(payload.get("lyric"), str) else ""
        return {"url": url, "cover": cover or "", "bitrate": DEFAULT_PLAY_LEVEL, "lyric": lyric}

    def _get_lyric_by_official_api(self, song_id: str) -> str:
        query = f"type=lyric&req=2&lrcx=1&rid={song_id}&songname=&artist=&corp=kuwo&fromchannel=bodian"
        encoded_query = base64.b64encode(query.encode("utf-8")).decode("utf-8")
        data = self._http.get_json(
            KUWO_LYRIC_URL,
            params={"f": "bodian", "q": encoded_query, "uid": "-1", "token": ""},
            timeout=REQUEST_TIMEOUT,
        )
        payload = data.get("data", {}) if isinstance(data, dict) else {}
        content = payload.get("content") if isinstance(payload, dict) else None
        if not isinstance(content, str) or not content:
            raise ValueError("Invalid bodian lyric")
        return base64.b64decode(content).decode("utf-8")

    def _get_by_tianbao(self, song_id: str) -> dict[str, str]:
        data = self._http.get_json(
            TIANBAO_API_URL,
            headers={"User-Agent": "Dart/2.19 (dart:io)", "plat": "ar", "channel": "aliopen"},
            params={
                "f": "web",
                "user": "2333333",
                "source": "kwplayerhd_ar_4.3.0.8_tianbao_T1A_qirui.apk",
                "type": "convert_url_with_sign",
                "br": DEFAULT_PLAY_BR,
                "rid": song_id,
            },
            timeout=REQUEST_TIMEOUT,
        )
        payload = data.get("data", {}) if isinstance(data, dict) else {}
        url = clean_text(str(payload.get("url", ""))) if isinstance(payload, dict) else ""
        if not is_http_url(url):
            raise ValueError("Invalid tianbao url")
        return {"url": url, "bitrate": DEFAULT_PLAY_BR}
